Terraform Policy (#237)
UKHO/docs: software-engineering-policies/InfrastructureAsCode/terraform.md

# Infrastructure as Code: Terraform

Terraform is our tool of choice, and should be considered in the first instance.

> Most standards and practises at the UKHO is covered in the Training courses we provide around the use of Terraform.

Reviewer: Add some more detail on this section which signposts who to contact if you need this training and where to find the pre-recorded videos.

> ## Avoid adding local modules

Reviewer 2: In this section could you spell out where you want them to go. I.E Modules should be stored in a github repo created for them.

Author: Think I have done that now?

Holding modules in the same repository as the project reduces it's audience. It is likely that you might be tempted to copy the modules you created in the previous project to your new one.

> This should be avoided at all costs. Most modules can be written in a relative IP lossless way, to avoid attack vertors or concerns around security, using public modules makes reuse easy to find and use when adhering to a standard naming makes them  easy to find in GitHub, using releases allows for ease of versioning.

Reviewer: ```suggestion
This should be avoided at all costs. Most modules can be written in a relative IP lossless way to avoid attack vectors or concerns around security. Using public modules makes them easy to find and reuse, adhering to a standard naming convention makes them easy to find in GitHub, and using releases allows for ease of versioning.
```

Author: applied in other update

> This should be avoided at all costs. Most modules can be written in a relative IP lossless way, to avoid attack vertors or concerns around security, using public modules makes reuse easy to find and use when adhering to a standard naming makes them  easy to find in GitHub, using releases allows for ease of versioning.

Reviewer 2: ```suggestion
This should be avoided. Most modules can be written in a relative IP lossless way, to avoid attack vertors or concerns around security, using public modules makes reuse easy to find and use when adhering to a standard naming makes them  easy to find in GitHub, using releases allows for ease of versioning.
```

Author: I think this crossed with some suggestions from el

A simple search for [tfmodule](https://github.com/UKHO?q=tfmodule&type=all&language=&sort=) will show a list of all the modules that are available and how to use them.

> ## Verisioning Modules

Reviewer: ```suggestion
## Versioning Modules
```

Author: applied in aother update

> Semantic versioning is used, the main branch is used to control the latest releases, when testing an addition, tagging the feature branch with a pre-release tag is a benefitial way to identify this. a main release will be versioned as 0.1.0 where as the pre-release will be versioned 0.1.0-alpha.1.

Reviewer: ```suggestion
Semantic versioning is used. The main branch is used to control the latest releases. When testing an addition, tagging the feature branch with a pre-release tag is a beneficial way to identify this. A main release will be versioned as 0.1.0 whereas the pre-release will be versioned 0.1.0-alpha.1.
```

> A new pre release will be needed for each change intended for testing.

Reviewer: ```suggestion
A new pre-release will be needed for each change intended for testing.
```

## Referencing modules in code

> Modules in github can be referenced easily using a source qualifier for github. the ref pointer in the source url refers to the tag / release version you intend to use.

Reviewer: ```suggestion
Modules in github can be referenced easily using a source qualifier for github. The ref pointer in the source url refers to the tag / release version you intend to use.
```

# Infrastructure as Code: Terraform

Terraform is our tool of choice, and should be considered in the first instance.

Most standards and practises at the UKHO is covered in the Training courses we provide around the use of Terraform.

> Videos and Presentation can be found in the **Software Engineering Team > Training Channel**.
> 
> Also online resources provided by [terraform.io](https://www.terraform.io/) are very useful.

Reviewer: ```suggestion
Videos and a presentation can be found in the **Software Engineering Team > Training Channel**. Online resources provided by [terraform.io](https://www.terraform.io/) are also very useful.
```

## Avoid adding local modules

Holding modules in the same repository as the project reduces it's audience. It is likely that you might be tempted to copy the modules you created in the previous project to your new one.

Most modules can be written in a relative IP lossless way, to avoid attack vectors or concerns around security. Using public modules makes them easy to find and reuse, adhering to a standard naming comvention makes them  easy to find in GitHub, and creating releases allows for ease of specifying a version to use.

> Each module should be hosted it's own GitHub repo to keep the download size small. following the convention mentioned in the [template project for terraform](https://github.com/UKHO/terraform-module-template) modules. Using the terraform module template will create a repo with the contents setup as standard. Review the naming convention guide linked on the readme of the template to see how you might go about naming the module.

Reviewer: ```suggestion
Each module should be hosted it's own GitHub repo to keep the download size small, following the convention mentioned in the [template project for terraform](https://github.com/UKHO/terraform-module-template) modules. Using the terraform module template will create a repo with the contents setup as standard. Review the naming convention guide linked on the readme of the template to see how you might go about naming the module.
```
Not sure I'm convinced by this? I get keeping the download size small but it feels like that's potentially a _lot_ of tiny repos, is there a better way?

> Using the same naming convention  allows for a [simple search for tfmodule](https://github.com/UKHO?q=tfmodule&type=all&language=&sort=) in GitHub to show a list of all the modules that are available and how to use them.

Reviewer: ```suggestion
Using the same naming convention allows for a [simple search for tfmodule](https://github.com/UKHO?q=tfmodule&type=all&language=&sort=) in GitHub to show a list of all the modules that are available and how to use them.
```

> ## Versioning Modules

Reviewer: ```suggestion
## Versioning modules
```

Semantic versioning is used. The main branch is used to control the latest releases. When testing an addition, tagging the feature branch with a pre-release tag is a beneficial way to identify this. A main release will be versioned as 0.1.0 whereas the pre-release will be versioned 0.1.0-alpha.1.

A new pre-release will be needed for each change intended for testing.

## Referencing modules in code

> Modules in github can be referenced easily using a source qualifier for github. The ref pointer in the source url refers to the tag / release version you intend to use.

Reviewer: ```suggestion
Modules in GitHub can be referenced easily using a source qualifier for GitHub. The ref pointer in the source url refers to the tag / release version you intend to use.
```
